spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/RemoveDanglingDeletesSparkAction.java

```diff
@@ -0,0 +1,227 @@
+            .select(
+                entries.col("partition"),
+                entries.col("spec_id"),
+                entries.col("file_path"),
+                entries.col("file_size_in_bytes"),
+                entries.col("record_count"));
+    MakeDeleteFile makePosDeleteFn =
+        new MakeDeleteFile(true, Partitioning.partitionType(table), table.specs());
+    Dataset<DeleteFile> posDeletesToRemove =
+        posDeleteDs.map(makePosDeleteFn, Encoders.javaSerialization(DeleteFile.class));
+
+    removedDeleteFiles.addAll(posDeletesToRemove.collectAsList());
+
+    // Dangling equality delete files
+    Dataset<Row> eqDeleteDs =
+        entries
+            .filter("content == 2") // equality delete files
+            .join(minDataSeqNumberPerPartition, joinCond)
+            .filter("sequence_number <= min_data_sequence_number")
+            .select(
+                entries.col("partition"),
+                entries.col("spec_id"),
+                entries.col("file_path"),
+                entries.col("file_size_in_bytes"),
+                entries.col("record_count"));
+    MakeDeleteFile makeEqDeleteFn =
+        new MakeDeleteFile(false, Partitioning.partitionType(table), table.specs());
+    Dataset<DeleteFile> eqDeletesToRemove =
+        eqDeleteDs.map(makeEqDeleteFn, Encoders.javaSerialization(DeleteFile.class));
+
+    removedDeleteFiles.addAll(eqDeletesToRemove.collectAsList());
+    return removedDeleteFiles;
+  }
+
+  private static class MakeDeleteFile implements MapFunction<Row, DeleteFile> {
+
+    private final boolean posDeletes;
+    private final Types.StructType partitionType;
+    private final Map<Integer, PartitionSpec> specsById;
+
+    /**
+     * Map function that transforms entries table rows into {@link DeleteFile}
+     *
+     * @param posDeletes true for position deletes, false for equality deletes
+     * @param partitionType partition type of table
+     * @param specsById table's partition specs
+     */
+    MakeDeleteFile(
+        boolean posDeletes, Types.StructType partitionType, Map<Integer, PartitionSpec> specsById) {
+      this.posDeletes = posDeletes;
+      this.partitionType = partitionType;
+      this.specsById = specsById;
+    }
+
+    @Override
+    public DeleteFile call(Row row) throws Exception {
+      PartitionData partition = new PartitionData(partitionType);
+      GenericRowWithSchema partitionRow = row.getAs(0);
+
+      for (int i = 0; i < partitionRow.length(); i++) {
+        partition.set(i, partitionRow.get(i));
+      }
+      int specId = row.getAs(1);
+      String path = row.getAs(2);
+      long fileSize = row.getAs(3);
+      long recordCount = row.getAs(4);
+
+      FileMetadata.Builder builder = FileMetadata.deleteFileBuilder(specsById.get(specId));
+      if (posDeletes) {
+        builder.ofPositionDeletes();
+      } else {
+        builder.ofEqualityDeletes();
+      }
+
+      return builder
+          .withPath(path)
+          .withPartition(partition)
```

Comment:
the partition data here doesn't comply with the builder spec, so it will build out incorrect partition data.

is it a bit of overwork to rebuild the deleteFile, cant we just delete using paths? a lot of overhead could be avoided.

Reply:
usually the copy needs to assert the specs are the same 

```
    public Builder copy(DeleteFile toCopy) {
      if (isPartitioned) {
        Preconditions.checkState(
            specId == toCopy.specId(), "Cannot copy a DeleteFile with a different spec");
        this.partitionData = DataFiles.copyPartitionData(spec, toCopy.partition(), partitionData);
      }
```

Reply:
it would be great if we can delete a deleteFile by filePaths only, but from mergingSnapshotProducer, looks like we never make it to work for path based deletion https://github.com/apache/iceberg/blob/d32abe8aa589e6c93cdc96dc7ae77ca5369af106/core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java#L193-L202